### ISSUE TITLE
Single instance single session TA fails to load second time

### ISSUE THREAD

**Issue by jforissier** [OP-TEE/optee_os#1918]
If a single instance, non-multiple-session TA does *not* call TEEC_CloseSession() before exiting, then it becomes impossible to run that TA again.

To reproduce:
- Modify `xtest` to make `--aes-perf` a single-instance TA (patch below)
- Run `xtest --aes-perf` twice.
- Expected: both runs should succeed
- What happens: second run reports: `TEEC_OpenSession: 0xffff000d (orig=3)`

Test on QEMU:
```
root@Vexpress:/ xtest --aes-perf
min=606.864us max=2711.76us mean=645.352us stddev=67.112us (cv 10.3993%) (1.51322MiB/s)
root@Vexpress:/ xtest --aes-perf
TEEC_OpenSession: 0xffff000d (orig=3)
```

`xtest` patch:
```diff
diff --git a/ta/aes_perf/include/user_ta_header_defines.h b/ta/aes_perf/include/user_ta_header_defines.h
index 54d8caf..d722b97 100644
--- a/ta/aes_perf/include/user_ta_header_defines.h
index 54d8caf..d722b97 100644
--- a/ta/aes_perf/include/user_ta_header_defines.h
+++ b/ta/aes_perf/include/user_ta_header_defines.h
@@ -34,7 +34,8 @@
 
 #define TA_FLAGS                (TA_FLAG_USER_MODE | TA_FLAG_EXEC_DDR | \
                                TA_FLAG_SECURE_DATA_PATH | \
-                               TA_FLAG_CACHE_MAINTENANCE)
+                               TA_FLAG_CACHE_MAINTENANCE | \
+                               TA_FLAG_SINGLE_INSTANCE)
 
 #define TA_STACK_SIZE          (2 * 1024)
 #define TA_DATA_SIZE           (32 * 1024)

```

**Comment by etienne-lms** [comment]
First analysis: this seems related to the preallocation of RPC command buffers in the TEE.

Optee pre-allocates the buffers used to send RPC commands. This prevents a SHM buffer allocation for each RPC request. Allocation is done from [`get_rpc_arg()`](https://github.com/OP-TEE/optee_os/blob/2.6.0/core/arch/arm/kernel/thread.c#L1284) and release from [`__thread_std_smc_entry()`](https://github.com/OP-TEE/optee_os/blob/2.6.0/core/arch/arm/kernel/thread.c#L600). If the "prealloc rpc cahe" is not disable, such SHM buffer is never freed. In current OP-TEE implementation, the feature is disabled only upon OP-TEE Linux driver unload... aka never.

So the first RPC request from an OP-TEE thread gets such OP-TEE thread specific buffer allocated and never released.

From OP-TEE Linux driver, all SHM allocations increment the caller TEE context refcount, see [`tee_shm_register()`](https://github.com/linaro-swg/linux/blob/optee-v4.9-20171027/drivers/tee/tee_shm.c#L278) and [`__tee_shm_alloc()`](https://github.com/linaro-swg/linux/blob/optee-v4.9-20171027/drivers/tee/tee_shm.c#L208). As a result, the each preallocated RPC command buffer increments the non-secure caller TEE context refcount. When the client legitimately closes its TEE context, the Linux driver does not release the realted context (and its resources) since Linux TEE driver still holds a refcount due to the allocated/unreleased buffer mention above.

This is how a Linux driver can keep a TEE context allocated even once it has terminate. If this TEE context is related to TEE resources (i.e a session) that are not explicitly released by the client, such a resource will never be released.

This applies to each optee thread (there are `CFG_NUM_THREADS` threads on TEE side): each can retain a shm buffer and hence a refcount on the at-the-time TEE context caller.

**Comment by jforissier** [comment]
@etienne-lms  thanks for looking into this.
Shouldn't the TEE driver take care of releasing all the resources associated with a client process on exit?

**Comment by etienne-lms** [comment]
That would more or less mean that RPC command buffers are freed each time a client terminate, and reallocated later. (note: this is only about the RPC command buffers, not any buffer allocated during RPC sequences).

I think the issue is in the relation between RPC command buffers (1 per OP-TEE thread) and the non-sequence clients. Maybe OP-TEE could ask the tee-supplicant to be owner of the RPC command buffers ? If supplicant crashed, Linux kernel should invoke the TEE the disable RPC buffer preallocation and release allocated buffers.

**Comment by lorc** [comment]
Maybe it is better to create separate context for that buffers?

**Comment by jenswi-linaro** [comment]
Or perhaps use no context at all.

**Comment by etienne-lms** [comment]
Anyway, this looks like a specific shm allocation scheme buffer TEE private ownership buffers.
=> A specific SHM allocation RPC command.

**Comment by jforissier** [comment]
> If the "prealloc rpc cahe" is not disable, such SHM buffer is never freed.

Confirmed, `CFG_PREALLOC_RPC_CACHE=n` is a workaround.

**Comment by etienne-lms** [comment]
`CFG_PREALLOC_RPC_CACHE` was introduced in release tag 3.14.0.
For prior OP-TEE versions, pick commit 8887663248ad ("core: new config switch CFG_PREALLOC_RPC_CACHE") and somewhat related fix 3cc421f72269 ("core: fix thread_disable_prealloc_rpc_cache()") that is recommended.

**Comment by jforissier** [comment]
Update: issue fixed in the Linux kernel driver. See https://lkml.org/lkml/2022/1/21/387.
